src/librustc/ty/layout.rs

```diff
@@ -1490,27 +1492,40 @@ impl<'gcx, 'tcx, T: HasTyCtxt<'gcx>> HasTyCtxt<'gcx> for LayoutCx<'tcx, T> {
     }
 }
 
-pub trait MaybeResult<T> {
+pub trait MaybeResult<T, E>: From<Result<T, E>> {
```

Comment:
So... apparently doing it this way is blocked on lazy normalization: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e6956340922d4df0087baa36b02b1ce3

Let's go with your original suggestion and add a `from(r: Result<T, Self::item>)` method to `MaybeResult`